skfem/assembly/basis/facet_basis.py

```diff
@@ -108,10 +115,102 @@ def default_parameters(self):
+            above for an example.
+        target_elem
+            Optional lower-dimensional finite element to project to.  If not
+            given, a piecewise constant element is used.
+
+        Returns
+        -------
+        InteriorBasis
+            An object corresponding to the trace mesh.
+        ndarray
+            A projected solution vector defined on the trace mesh.
+
+        """
+        from skfem.utils import project
+
+        DEFAULT_TARGET = {
+            MeshTri: ElementLineP0,
+            MeshQuad: ElementLineP0,
+            MeshTet: ElementTriP0,
+            MeshHex: ElementQuad0,
+        }
+
+        meshcls = type(self.mesh)
+        if meshcls not in DEFAULT_TARGET:
+            raise NotImplementedError("Mesh type not supported.")
+        if target_elem is None:
+            target_elem = DEFAULT_TARGET[meshcls]()
+
+        TRACE_RESTRICT_MAP: Dict[Tuple[Type[Element], Type[Mesh]], Any] = {
+            (ElementLineP0, MeshTri): (ElementTriP0, MeshLine),
+            (ElementLineP1, MeshTri): (ElementTriP1, MeshLine),
+            (ElementLineP2, MeshTri): (ElementTriP2, MeshLine),
+            (ElementLineP0, MeshQuad): (ElementQuad0, MeshLine),
+            (ElementLineP1, MeshQuad): (ElementQuad1, MeshLine),
+            (ElementLineP2, MeshQuad): (ElementQuad2, MeshLine),
+            (ElementTriP0, MeshTet): (ElementTetP0, MeshTri),
+            (ElementTriP1, MeshTet): (ElementTetP1, MeshTri),
+            (ElementTriP2, MeshTet): (ElementTetP2, MeshTri),
+            (ElementQuad0, MeshHex): (ElementHex0, MeshQuad),
+            (ElementQuad1, MeshHex): (ElementHex1, MeshQuad),
+            (ElementQuad2, MeshHex): (ElementHex2, MeshQuad),
+        }
+
+        if (type(target_elem), meshcls) not in TRACE_RESTRICT_MAP:
+            raise Exception("The specified 'elem' not supported.")
+        elemcls, target_meshcls = TRACE_RESTRICT_MAP[(type(target_elem),
+                                                      meshcls)]
+
+        fbasis = FacetBasis(self.mesh,
+                            elemcls(),
+                            facets=self.find,
+                            quadrature=(self.X, self.W))
+        I = fbasis.get_dofs(self.find).all()
+        if len(I) == 0:  # special case: no facet DOFs
+            if fbasis.dofs.interior_dofs.shape[0] > 1:
+                # no one-to-one restriction: requires interpolation
+                raise NotImplementedError
+            # special case: piecewise constant elem
+            I = fbasis.dofs.interior_dofs[:, self.tind].flatten()
+        y = project(x, self, fbasis, I=I)
+
+        # build connectivity for lower dimensional mesh
+        ix = self.mesh.facets[:, self.find]
+        ixuniq = np.unique(ix)
+        b = np.zeros(np.max(ix) + 1, dtype=np.int64)
+        b[ixuniq] = np.arange(len(ixuniq), dtype=np.int64)
+
+        return InteriorBasis(
+            target_meshcls(projection(self.mesh.p[:, ixuniq]), b[ix]),
```

Comment:
Now the questions remains: is this sufficient or should we support more general use cases, e.g., sometimes the user might actually not want to initialize `InteriorBasis` or even `target_meshcls` but simply feed `self.mesh.p[:, ixuniq]` and `b[ix]` to some external tool together with `y`.

Reply:
One solution would be to split everything else but the return statement into a separate method which can be called for different use cases. I'll try that.

Reply:
Now it should be flexible enough for different use cases (by direct use of `FacetBasis._trace`) and reasonable defaults (in `FacetBasis.trace`).